### PR TITLE
[Enhancement] QueryCache support primary_key and unique_key

### PR DESCRIPTION
**Pull request by satanson**
## What type of PR is this：
- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes ：

Fixes #

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

Query cache support data models of  UNIQUE_KEYS and PRIMARY_KEYS. now query cache support all data models. different models have different policies to support multiversion:
1. support multiversion cache:  DUP_KEYS, AGG_KEYS(contains no replace aggfunc)
2. not support multiversion cache: UNIQUE_KEYS, PRIMARY_KEYS, AGG(contains replace aggfunc). Althrough these scenarios do not support multiversion cache, but if cache contains a stale cache value and the delta version has not data, the cache can be hit totally.

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 68 / 72 (94.44%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/planner/OlapScanNode.java|37|40|92.50%|[534, 535, 723]|
|:large_blue_circle:|com/starrocks/planner/FragmentNormalizer.java|18|19|94.74%|[116]|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|8|8|100.00%|[]|
|:large_blue_circle:|com/starrocks/planner/PlanFragment.java|5|5|100.00%|[]|

File: fe/fe-core/src/main/java/com/starrocks/planner/OlapScanNode.java [hunk 1]
```diff
@@ -751,33 +782,43 @@ public void normalizeConjuncts(FragmentNormalizer normalizer, TNormalPlanNode pl
...
+            case UNIQUE_KEYS:
+                normalizer.setCanUseMultiVersion(false);
+                return false;
             case AGG_KEYS: {
                 List<Column> columns = selectedIndexId == -1 ? olapTable.getBaseSchema() :
                         olapTable.getSchemaByIndexId(selectedIndexId);
-                return columns.stream().anyMatch(
+                boolean hasReplaceAgg = columns.stream().anyMatch(
                         c -> c.isAggregated() && c.getAggregationType().isReplaceFamily());
+                normalizer.setCanUseMultiVersion(!hasReplaceAgg);
```
Comment (LiShuMing) [inline fe/fe-core/src/main/java/com/starrocks/planner/OlapScanNode.java:812]: What's the behavior when the cache is uncacheable and canUseMultiVersion?

File: fe/fe-core/src/main/java/com/starrocks/qe/SessionVariable.java [hunk 2]
```diff
@@ -739,6 +744,12 @@ public boolean getUseScanBlockCache() {
     @VarAttr(name = QUERY_CACHE_ENTRY_MAX_ROWS)
     private long queryCacheEntryMaxRows = 409600;
 
+    @VarAttr(name = QUERY_CACHE_PRIMARY_KEYS_HOT_PARTITION_NUM)
+    private int queryCachePrimaryKeysHotPartitionNum = 3;
```
Comment (LiShuMing) [inline fe/fe-core/src/main/java/com/starrocks/qe/SessionVariable.java:748]: Add some documents for these two variables?

File: fe/fe-core/src/main/java/com/starrocks/planner/OlapScanNode.java [hunk 3]
```diff
@@ -707,6 +715,26 @@ public long getSelectedIndexId() {
...
+        if (connectContext == null) {
+            return hotIds;
+        }
+        List<Map.Entry<Long, Range<PartitionKey>>> partitions = partitionInfo.getSortedRangeMap(false);
+        int numHotIds = 0;
+        int numPartitions = partitions.size();
+        if (keysType == KeysType.PRIMARY_KEYS) {
+            int pkHotNum = connectContext.getSessionVariable().getQueryCachePrimaryKeysHotPartitionNum();
+            numHotIds = Math.min(numPartitions, Math.max(0, pkHotNum));
+        } else if (keysType == KeysType.UNIQUE_KEYS || (keysType == KeysType.AGG_KEYS && !canUseMultiversion)) {
```
Comment (LiShuMing) [inline fe/fe-core/src/main/java/com/starrocks/planner/OlapScanNode.java:731]: Why AGG_KEYS && ! canUseMultiversion needs this? is it uncacheable?

Comment (LiShuMing) [reply fe/fe-core/src/main/java/com/starrocks/planner/OlapScanNode.java:731]: Can we just use canUseMultiversion without depending on KeysType?

File: be/src/exec/query_cache/cache_operator.cpp [hunk 4]
```diff
@@ -268,6 +278,40 @@ void CacheOperator::_handle_stale_cache_value(int64_t tablet_id, CacheValue& cac
...
+    }
+    auto& [tablet, rowsets] = status.value();
+    const auto snapshot_version = cache_value.version;
+    bool has_compacted_delta_rowsets = false;
+    bool exists_non_empty_delta_rowsets = false;
+    for (auto& rs : rowsets) {
+        has_compacted_delta_rowsets |= rs->start_version() == snapshot_version + 1;
+        exists_non_empty_delta_rowsets |= rs->start_version() > snapshot_version && rs->has_data_files();
+    }
+    if (exists_non_empty_delta_rowsets && has_compacted_delta_rowsets) {
```
Comment (LiShuMing) [inline be/src/exec/query_cache/cache_operator.cpp:301]: Why  `has_compacted_delta_rowsets `  affects? why not `exists_non_empty_delta_rowsets || exists_non_empty_delta_rowsets `?

File: be/src/exec/query_cache/cache_operator.cpp [hunk 5]
```diff
@@ -237,8 +246,9 @@ void CacheOperator::_handle_stale_cache_value(int64_t tablet_id, CacheValue& cac
     Version delta_versions(min_version, max_version);
     buffer->tablet = tablet;
     auto has_delete_predicates = tablet->has_delete_predicates(delta_versions);
-    // case 1: there exist delete predicates in delta versions, the cache result is not reuse, so cache miss.
-    if (has_delete_predicates) {
+    // case 1: there exist delete predicates in delta versions, or data model can not support multiversion cache and
+    // the tablet has non-empty delta rowsets; then cache result is not reuse, so cache miss.
+    if (has_delete_predicates || (!_cache_param.can_use_multiversion && !all_rs_empty)) {
```
Comment (LiShuMing) [inline be/src/exec/query_cache/cache_operator.cpp:251]: `has_delete_predicates ` is supported when `all_rs_empty `  is true?

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 68 / 72 (94.44%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/planner/OlapScanNode.java|37|40|92.50%|[534, 535, 723]|
|:large_blue_circle:|com/starrocks/planner/FragmentNormalizer.java|18|19|94.74%|[116]|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|8|8|100.00%|[]|
|:large_blue_circle:|com/starrocks/planner/PlanFragment.java|5|5|100.00%|[]|

File: fe/fe-core/src/main/java/com/starrocks/planner/OlapScanNode.java (hunk 1 above)
Comment (satanson) [reply fe/fe-core/src/main/java/com/starrocks/planner/OlapScanNode.java:812]: uncacheable is confusing,  so it is renamed to setIfCanUseMultiVersionCache. the behavior is commented

File: fe/fe-core/src/main/java/com/starrocks/qe/SessionVariable.java (hunk 2 above)
Comment (satanson) [reply fe/fe-core/src/main/java/com/starrocks/qe/SessionVariable.java:761]: Comment has been already added at line 299~300

File: fe/fe-core/src/main/java/com/starrocks/planner/OlapScanNode.java (hunk 3 above)
Comment (satanson) [reply fe/fe-core/src/main/java/com/starrocks/planner/OlapScanNode.java:731]: No, we can not, primary key and unique key can not support multiversion cache, to see the comment.

File: be/src/exec/query_cache/cache_operator.cpp (hunk 5 above)
Comment (satanson) [reply be/src/exec/query_cache/cache_operator.cpp:251]: It is just two cases can be handled in cache-miss mode, to see the comment

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 65 / 69 (94.20%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/planner/OlapScanNode.java|34|37|91.89%|[534, 535, 723]|
|:large_blue_circle:|com/starrocks/planner/FragmentNormalizer.java|18|19|94.74%|[116]|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|8|8|100.00%|[]|
|:large_blue_circle:|com/starrocks/planner/PlanFragment.java|5|5|100.00%|[]|

File: be/src/exec/query_cache/cache_operator.cpp (hunk 4 above)
Comment (satanson) [reply be/src/exec/query_cache/cache_operator.cpp]: the condition is wrong, fixed

**Review by kangkaisen — APPROVED** [review]

**Comment by satanson** [comment]
run starrocks_fe_unittest

**Review by stdpain — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Review by ZiheLiu — APPROVED** [review]

**Comment by satanson** [comment]
run starrocks_fe_unittest

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 56 / 62 (90.32%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/planner/OlapScanNode.java|30|34|88.24%|[534, 535, 723, 811]|
|:large_blue_circle:|com/starrocks/planner/FragmentNormalizer.java|17|19|89.47%|[108, 116]|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/planner/PlanFragment.java|5|5|100.00%|[]|

**Comment by satanson** [comment]
@mergifyio backport branch-2.5

**Pull request merged**